logcheck/pkg/logcheck.go

```diff
@@ -182,37 +182,43 @@ func checkForFunctionExpr(fexpr *ast.CallExpr, pass *analysis.Pass, c *config) {
 				return
 			}
 
-			if keyCheckEnabled || parametersCheckEnabled || valueCheckEnabled {
-				// if format specifier is used, check for arg length will most probably fail
-				// so check for format specifier first and skip if found
-				if parametersCheckEnabled && checkForFormatSpecifier(fexpr, pass) {
-					return
-				}
-				switch fName {
-				case "InfoS":
-					kvCheck(args[1:], fun, pass, fName, keyCheckEnabled, parametersCheckEnabled, valueCheckEnabled)
-				case "ErrorS":
-					kvCheck(args[2:], fun, pass, fName, keyCheckEnabled, parametersCheckEnabled, valueCheckEnabled)
+			// variadic input is a valid input to klog.Error*, klog.Info*, logr.Logger.Info and logr.Logger.Error
+			// functions. Hence checking the parameters for variadic input argument is excluded.
+			if !fexpr.Ellipsis.IsValid() {
```

Comment:
This solves the problem, but also opens the door for indirect bugs where wrapper functions around logging calls are called incorrectly.

Would it be possible to do something similar as for fmt.Sprintf and detect such wrapper functions? Then when those get called, logcheck may have access to the values and can check them. That should work at least for simple cases:

```
func foo() {
   myWrapper("hello", "key1", "value1", 1, 2) // Not okay! 1 is not a string.
}

func myWrapper(msg string, args ...interface{}) { klog.InfoS(msg, args...) }
```

I've never used the mechanism myself, but I know that it is possible to annotate items and later retrieve that information.

Reply:
It is true that this fix will open the indirect bug. I think at the static-checking level it is particularly difficult to check the values inside the variadic argument. What do you suggest to improve this test?

Reply:
When a function or method "foo" calls a log call, tag that function. Then whenever *any* function or method is called, check that tag and validate it's parameters accordingly. I don't know how to do this, but I am pretty sure that it is possible.

Do you want to explore this? I'm sure it's going to be a fun exercise, but not trivial.

If not, then we can also merge this PR as-is and open an enhancement issue for the remaining work.

Reply:
It sounds interesting to me and would like to explore it.

I would propose a new PR for above mentioned improvement.

Reply:
Okay, let's do that. However, some tests are currently failing.